src/datasets/features/image.py

```diff
@@ -0,0 +1,287 @@
+                raise IndexError("Invalid take for empty PandasImageExtensionArray, must be all -1.")
+            else:
+                data = np.array([fill_value] * len(indices), dtype=np.object)
+                return PandasImageExtensionArray(data, copy=False)
+        took = self._data.take(indices)
+        if allow_fill and mask.any():
+            took[mask] = [fill_value] * np.sum(mask)
+        return PandasImageExtensionArray(took, copy=False)
+
+    def map(self, mapper):
+        # More info about this (undocumented) function can be found here:
+        # https://github.com/pandas-dev/pandas/issues/23179
+        return PandasImageExtensionArray(pd.Series(self._data).map(mapper).to_numpy())
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+    def __eq__(self, other) -> np.ndarray:
+        if not isinstance(other, PandasImageExtensionArray):
+            raise NotImplementedError(f"Invalid type to compare to: {type(other)}")
+        return (self._data == other._data).all()
+
+
+@dataclass(unsafe_hash=True)
+class Image:
+    """Image feature to read image data from an image file.
+
+    Input: The Image feature accepts as input:
+    - A :obj:`str`: Absolute path to the image file (i.e. random access is allowed).
+    - A :obj:`dict` with the keys:
+
+        - path: String with relative path of the image file to the archive file.
+        - bytes: Bytes of the image file.
+
+      This is useful for archived files with sequential access.
+
+    - An :obj:`np.ndarray`: NumPy array representing an image.
+    - A :obj:`PIL.Image.Image`: PIL image object.
+    """
+
+    id: Optional[str] = None
+    # Automatically constructed
+    dtype: ClassVar[str] = "dict"
+    pa_type: ClassVar[Any] = None
+    _type: str = field(default="Image", init=False, repr=False)
+
+    def __call__(self):
+        return ImageExtensionType()
+
+    def encode_example(self, value):
+        """Encode example into a format for Arrow.
+
+        Args:
+            value (:obj:`str`, :obj:`np.ndarray`, :obj:`PIL.Image.Image` or :obj:`dict`): Data passed as input to Image feature.
+
+        Returns:
+            :obj:`dict`
+        """
+        if config.PIL_AVAILABLE:
+            import PIL.Image
+        else:
+            raise ImportError("To support encoding images, please install 'Pillow'.")
+
+        if isinstance(value, str):
+            return {"path": value, "bytes": None}
+        elif isinstance(value, np.ndarray):
+            image = PIL.Image.fromarray(value.astype(np.uint8))
+            return {"path": None, "bytes": image_to_bytes(image)}
+        elif isinstance(value, PIL.Image.Image):
+            if hasattr(value, "filename") and value.filename != "":
+                return {"path": value.filename, "bytes": None}
+            else:
+                return {"path": None, "bytes": image_to_bytes(value)}
+        else:
+            return value
+
+    def decode_example(self, value):
```

Comment:
```suggestion
    def decode_example(self, value: dict) -> "PIL.Image.Image":
```

Reply:
We also support strings as inputs to allow `cast_column` calls on the columns that store image file paths. Let me know what do you think about that. If you are ok with that option, then we should have:
```suggestion
    def decode_example(self, value: Union[str, dict]) -> "PIL.Image.Image":
```